sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/unresolved.scala

```diff
@@ -276,6 +279,138 @@ object UnresolvedAttribute {
+ */
+case class UnresolvedAttributeIdentifierClause(expr: Expression)
+  extends Expression with Unevaluable {
+
+  override def children: Seq[Expression] = Seq(expr)
+
+  override def dataType: DataType = throw new UnresolvedException("dataType")
+  override def nullable: Boolean = throw new UnresolvedException("nullable")
+  override lazy val resolved = false
+  final override val nodePatterns: Seq[TreePattern] = Seq(UNRESOLVED_IDENTIFIER_CLAUSE)
+
+  override def prettyName: String = "IDENTIFIER"
+  override def toString: String = {
+    s"'(${children.mkString(", ")})"
+  }
+
+  override protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]):
+  UnresolvedAttributeIdentifierClause = {
+      copy(expr = newChildren.head)
+  }
+}
+
+case class UnresolvedFunctionIdentifierClause(
+                               identExpr: Expression,
+                               arguments: Seq[Expression],
+                               isDistinct: Boolean,
+                               filter: Option[Expression] = None,
+                               ignoreNulls: Boolean = false)
+  extends Expression with Unevaluable {
+
+  override def children: Seq[Expression] = arguments ++ filter.toSeq
+
+  override def dataType: DataType = throw new UnresolvedException("dataType")
+  override def nullable: Boolean = throw new UnresolvedException("nullable")
+  override lazy val resolved = false
+  final override val nodePatterns: Seq[TreePattern] = Seq(UNRESOLVED_IDENTIFIER_CLAUSE)
+
+  override def prettyName: String = identExpr.toString
+  override def toString: String = {
+    val distinct = if (isDistinct) "distinct " else ""
+    s"'${identExpr.toString}($distinct${children.mkString(", ")})"
+  }
+
+  override protected def withNewChildrenInternal
+  (newChildren: IndexedSeq[Expression]): UnresolvedFunctionIdentifierClause = {
+    if (filter.isDefined) {
+      copy(arguments = newChildren.dropRight(1), filter = Some(newChildren.last))
+    } else {
+      copy(arguments = newChildren)
+    }
+  }
+}
+
+object UnresolvedFunctionIdentifierClause {
+  def apply(
+             identExpr: Expression,
+             arguments: Seq[Expression],
+             isDistinct: Boolean): UnresolvedFunction = {
+    UnresolvedFunctionIdentifierClause(identExpr, arguments, isDistinct)
+  }
+}
+
+/**
+ * A table-valued function, e.g.
+ * {{{
+ *   select id from range(10);
+ * }}}
+ *
+ * @param identExpr user-specified expression with teh name of this table-value function
+ * @param functionArgs list of function arguments
+ */
+case class UnresolvedTableValuedFunctionIdentifierClause(
+                                          identExpr: Expression,
+                                          functionArgs: Seq[Expression])
+  extends LeafNode {
+
+  override def output: Seq[Attribute] = Nil
+
+  override lazy val resolved = false
+
+  final override val nodePatterns: Seq[TreePattern] = Seq(UNRESOLVED_IDENTIFIER_CLAUSE)
```

Comment:
optional: do we want to create separate patterns for each of these types of unresolved identifier clause? Or just combine them all into one? Having separate ones could grant more flexibility later if needed?

Reply:
I only use these for the query ones by rote... I suppose we can change them at will.